pkg/agent/validations.go

```diff
@@ -2,138 +2,123 @@ package agent
+		return nil, errors.Wrap(err, "unable to verify validations")
 	}
 
-	for category, validationResults := range validationsInfo {
+	for _, validationResults := range validationsInfo {
 		for _, r := range validationResults {
+			// If validation ID does not exist create it
+			if _, ok := validationHistory[r.ID]; !ok {
+				validationHistory[r.ID] = &validationResultHistory{
+					numFailures:     0,
+					seen:            false,
+					currentStatus:   "",
+					currentMessage:  "",
+					previousStatus:  "",
+					previousMessage: "",
+				}
+			}
+			// First time we see a validation ID
+			if !validationHistory[r.ID].seen {
+				validationHistory[r.ID].currentMessage = r.Message
+				validationHistory[r.ID].currentStatus = r.Status
+			} else {
+				validationHistory[r.ID].previousMessage = validationHistory[r.ID].currentMessage
+				validationHistory[r.ID].previousStatus = validationHistory[r.ID].currentStatus
+				validationHistory[r.ID].currentMessage = r.Message
+				validationHistory[r.ID].currentStatus = r.Status
+			}
 			switch r.Status {
 			case validationFailure, validationError:
-				label := r.ID
-				if v, ok := hostValidationLabels[r.ID]; ok {
-					label = v
-				}
-
-				traces = append(traces, validationTrace{
-					header:   errorMsg,
-					category: category,
-					label:    label,
-					message:  r.Message,
-				})
+				validationHistory[r.ID].numFailures++
 			}
+			logValidationHistory(logPrefix, validationHistory[r.ID], log)
 		}
 	}
-
-	return traces
+	return validationHistory, nil
 }
 
-func checkHostsValidations(cluster *models.Cluster, log *logrus.Logger) bool {
-
-	var currentValidations []validationTrace
-
-	currentValidations = append(currentValidations, logValidationsStatus("Validation failure found for cluster", cluster.ValidationsInfo, log)...)
-	for _, h := range cluster.Hosts {
-		currentValidations = append(currentValidations, logValidationsStatus(fmt.Sprintf("Validation failure found for %s", h.RequestedHostname), h.ValidationsInfo, log)...)
-	}
-
-	sort.Slice(currentValidations, func(i, j int) bool {
-		if currentValidations[i].header != currentValidations[j].header {
-			return currentValidations[i].header < currentValidations[j].header
-		}
-		if currentValidations[i].category != currentValidations[j].category {
-			return currentValidations[i].category < currentValidations[j].category
-		}
-		return currentValidations[i].label < currentValidations[j].label
-	})
-
-	if !reflect.DeepEqual(currentValidations, previousValidations) {
-		previousValidations = currentValidations
-
-		if len(previousValidations) == 0 {
-			log.Info("Pre-installation validations are OK")
-			return true
+func logValidationHistory(logPrefix string, history *validationResultHistory, log *logrus.Logger) string {
+	loglevel := ""
+	// Fist time we print anything
+	if !history.seen {
+		switch history.currentStatus {
+		case validationFailure, validationError:
```

Comment:
This is a duplicate of the code below at line 111. That could be eliminated by only returning early in the success case:

```
if !history.seen {
    history.seen = true
    if history.currentStatus = validationSuccess {
        log.Debug(...)
        return
    }
}
```

Reply:
There is actually a [decent amount of statuses in assisted-service](https://github.com/openshift/assisted-service/blob/dac620ca40898913e5c847424aec5dc1b84d1a78/internal/host/validator.go#L40-L45) we don't act on. Does that change the logic here?